src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -278,5 +278,76 @@ public static List<string> LookupNames(this SemanticModel model, int position, I
                     : model.LookupSymbols(position, container);
             return symbols.Select(s => s.Name).Distinct().ToList();
         }
+
+        internal static void VerifyTypes(this CSharpCompilation compilation)
+        {
+            foreach (var tree in compilation.SyntaxTrees)
+            {
+                VerifyTypes(compilation, tree);
+            }
+        }
+
+        internal static void VerifyTypes(this CSharpCompilation compilation, SyntaxTree tree)
+        {
+            var root = tree.GetRoot();
+            var annotations = GetAnnotations(root);
+            if (annotations.IsEmpty)
+            {
+                return;
+            }
+            var model = compilation.GetSemanticModel(tree);
+            var expectedTypes = annotations.SelectAsArray(annotation => annotation.Text);
+            var actualTypes = annotations.SelectAsArray(
+                annotation => ((CSharpDataFlowAnalysis)model.AnalyzeDataFlow(annotation.Expression)).TypeAndNullability.ToDisplayString(TypeSymbolWithAnnotations.DebuggerDisplayFormat));
+            // Consider reporting the correct source with annotations on mismatch.
+            AssertEx.Equal(expectedTypes, actualTypes);
+        }
+
+        private static ImmutableArray<(ExpressionSyntax Expression, string Text)> GetAnnotations(SyntaxNode root)
```

Comment:
>private static ImmutableArray<(ExpressionSyntax Expression, string Text)> GetAnnotations(SyntaxNode root) [](start = 8, length = 105)

Can we make this a local function to avoid polluting the "namespace"?